text/0000-return-type-notation.md

```diff
@@ -0,0 +1,1113 @@
+    //  --------------- stricter than necessary 
+{
+    tokio::spawn(async move {
+        server.shutdown().await;
+    });
+}
+```
+
+Because `spawn_shutdown` only invokes `shutdown`, using `SendHealthCheck` is stricter than necessary.
+It may be that there are types where the `check` method does not return a `Send` future
+but `shutdown` does.
+In this case, you can write a bound that specifically applies to the future returned by the `shutdown()` method, like so:
+
+```rust
+fn spawn_shutdown<HC>(health_check: H, server: Server)
+where
+    HC: HealthCheck<shutdown(..): Send> + Send + 'static,
+    //              ------------------ "just right"
+{
+    tokio::spawn(async move {
+        server.shutdown().await;
+    });
+}
+```
+
+The `shutdown(..)` notation acts like an associated type referring to the return type of the method.
+The bound `HC: HealthCheck<shutdown(..): Send>` indicates that the `shutdown` method,
+regardless of what arguments it is given,
+will return a `Send` future. 
+These bounds do not have to be written in the `HealthCheck` trait, it could also be written as follows:
+
+```rust
+fn spawn_shutdown<HC>(health_check: H, server: Server)
+where
+    HC: HealthCheck + Send + 'static,
+    HC::shutdown(..): Send,
+```
+
+## Guidelines and best practices
+
+### Authoring async traits
+
+When defining an async trait (a trait with async functions), best practice is to define a "send variant" with the `trait_variant` crate:
+
+```rust
+#[trait_variant::make(SendMyTrait: Send)]
+trait MyTrait {
+    async fn method1(&self);
+    async fn method2(&self);
+}
+```
+
+Defining a "send alias" in this way has advantages for users of your trait:
+
+* Referencing `T: SendMyTrait` is shorter than using RTN if there are multiple functions
+    * (compare to `T: Send + Mytrait<method1(..): Send, method2(..): Send>`)
+* Referencing `T: SendMyTrait` is more forwards compatible:
+    * If you add a new method to your trait (with a default impl), all users of the send alias will be able to call this new method. Users that have named individual methods will not (on the flip side)
+
+But defining a "send alias" in this way comes with obligations for you:
+
+* If you add a new default method to your trait, it must be "Send-preserving" (meaning that it will be `Send` if other functions return `Send` futures).
+    * *Why?* If there is an existing function that requires `T: SendMyTrait` for some type `T`, then this must remain true even when `MyTrait` grows a new (defaulted) method, or else you will have broken your downstream clients.
+    * On the flip side, if you don't define an alias, you can add new defaulted methods that are not Send. This won't break downstream crates but neither will they be able to use them.
+
+### Using async traits
+
+When using a trait `MyTrait` that defines a sendable alias `SendMyTrait`...
+
+* Implement `MyTrait` directly. Your type will implement `SendMyTrait` automatically if appropriate.
```

Comment:
Interesting.. I had to think about whether I agree with this and started to argue against it, but I think it's probably the right call.

Implementing `SendMyTrait` directly can be nice for concrete types, because makes it your contracts more explicit. Of course this is a special case of auto trait leakage, but here we have the opportunity to mitigate its downsides with best practices. But a big problem with this take is that the converse is not true; implementing `MyTrait` does not remove any semver hazards from your code, it only hides them. I would not want to lead users into a false sense of security here.

The other problem is that the practice of implementing `SendMyTrait` doesn't extend to generic types. You could argue that the resulting need to add a `where T: Send` or `where T: SendMyTrait` to `impl<T> SendMyTrait for Struct<T>` might cause an implementer to think twice about whether their impl needs to be so constrained. But then the key question becomes whether they _know_ they can write `impl<T> MyTrait for Struct<T>` and still have `SendMyTrait` implemented when possible. In the case where you always implement `MyTrait` directly, this quickly becomes obvious from experience.

With all that said, I might personally choose to implement `SendMyTrait` directly in cases where it's important.. but that would be going against the general best practices laid out here (which I agree with), and is probably best left for the cases where I would otherwise bother to write a static assertion.

Note that this is connected to the convention of whether to use `MyTrait`/`SendMyTrait` or `LocalMyTrait`/`MyTrait`. In the latter case it would seem we are encouraging impls to use the send variant by default and to "opt in" to more genericity / fewer guarantees.

Reply:
Good points. I think when I wrote this, I was also thinking about the fact that you cannot implement trait aliases today. I think we should expedite work on changing that, in which case I suspect that the best advice would be to implement the Send variant if that's what you always want.

Reply:
I'm not so concerned about people learning that they can implement the 'non-send' variant and still get the benefits. They'll get that in time, I think, I don't think they have to learn it right away. Using the send variant will get you faster and better error messages.